Adjustments

kristajg opened the pull request:
Now up to date with master.  There were some pretty nasty merge conflicts, but I think I ironed everything out (but keep in mind when reviewing to be sure it's all intact).

- Table is fully styled
- Added helper function for 'chunking' an array of data for styling purposes... styling dynamic content without bootstrap or flexbox is a little hard without doing this method

NOTE: Should I squash commits now, or previous to merge?

brad-decker commented on src/components/masthead/Delete.js:
```diff
@@ -2,15 +2,30 @@ import React from 'react';
-    <DeleteIcon style={iconStyle} onClick={handleDelete(itemsSelected)} />
-  </ListItem>
-);
+const Delete = ({ itemsSelected, handleDelete }) => {
```
You need to switch out the curly brackets here for () so that the component is returned. Currently getting an error when trying to click checkboxes after building the package.

kristajg replied:
Ok, that is currently in master right now from the last merge

Nathan-Schwartz replied:
It looks to me like master has parenthesis

brad-decker commented on src/components/masthead/Delete.js:
```diff
@@ -2,15 +2,30 @@ import React from 'react';
+    {`${itemsSelected.length} item${itemsSelected.length <= 1 ? '' : 's'} selected`}
+    <DeleteIcon
+      style={styles.iconStyle}
+      onClick={handleDelete(itemsSelected)} />
```
This is actually calling handleDelete on every render call. The simplest but incorrect solution would be to just wrap it in an arrow function or even a .bind statement. The reason we don't do these is because creating new functions in render methods is horrible for performance. The functions get created, and then jettisoned - but if they are referenced from other components that aren't unmounted they stay in the memory heap of the browser. 

Ideally what i'd suggest is that this component only receives the current count of items selected, and a handleDelete method that is simply a pointer to a function on the root component where itemsSelected is actually being managed (in this case, list/index.js). So in listIndex you might have something like:

```js
handleDelete() {
  this.props.handleDelete(this.state.selectedItems);
}
```

then pass that instead of this.props.handleDelete to your child components so that it gets curried down to this Delete component.

Nathan-Schwartz replied:
I think this one is one me, but I agree with the feedback 👍

kristajg replied:
@brad-decker This is good feedback, thanks.  I'll go with the ideal approach

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -2,21 +2,19 @@ import React, { Component } from 'react';
         return acc;
       }, {}),
     };
+    this.toggleVisibility = this.toggleVisibility.bind(this);
```
in general binding like this is okay in constructor method, but a better solution is to rely on es6 class properties to help us with this in combination of arrow methods. 

for example i can rewrite this component without a constructor method using class properties. 

```js
class Filter extends Component {
  state = {
    opsVisible: false,
    ...this.props.filters.reduce((acc, filter) => {
      acc[filter.label] = [];
      return acc;
    }, {}),
  };

  toggleVisibility = () => this.setState({ opsVisible: !this.state.opsVisible });
```

by using that arrow function this will always be scoped to the appropriate context without having to bind. es6 class properties that are assignments are evaluated in the constructor call automagically. The result is a nice clean syntax.

kristajg replied:
Awesome, this way is much better

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -43,29 +41,69 @@ class Filter extends Component {
+
   render() {
+    const { filters } = this.props;
+    const filterColumns = chunkArray(filters, 4);
```
I think we should utilize reselect's createSelector to wrap chunkArray to prevent having to compute the columns on every render. If filters doesn't change it'll just return the cached result which will be clutch when this is added into more expensive apps.

brad-decker replied:
We could literally add that selector to the helper file so this file doesn't even have to change really, but wanted to point it out here too :)

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -43,29 +41,69 @@ class Filter extends Component {
+    const styles = {
+      filterBar: {
+        color: 'rgb(3, 169, 244)',
+        fontSize: '14px',
```
for all styles that are just pixel values (with the exception of lineHeight) we prefer just numeric values instead of strings with units. so for example

```js
filterBar: {
  fontSize: 14,
}
```

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -43,29 +41,69 @@ class Filter extends Component {
+        fontSize: '14px',
+        fontWeight: 'regular',
+        borderBottom: '1px solid rgb(224, 224, 224)',
+        height: '200px',
```
Remove height forced to 200px. We will instead use a clear: both style object to help us with the floated items. add a new key to styles like this:

```js
floatClear: {
  display: 'block',
  clear: 'both',
};
```

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -43,29 +41,69 @@ class Filter extends Component {
-            </ListItem>
-          ),
-        ]} />
+            </ul>
```
after this </ul> add:
```js
<span style={styles.floatClear} />
```

kristajg replied:
@brad-decker What is the desired corresponding style for floatClear?  clear: 'both'?

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -43,29 +41,69 @@ class Filter extends Component {
+          onClick={this.toggleVisibility} />
+        <ListItem
+          style={{ ...styles.filterOptions, display: this.state.opsVisible ? 'block' : 'none' }}>
+          {filterColumns.map((col, i) =>
```
One short coming that I see here in regards to the design is that this doesn't take into account the possibility of having multiple fields we might want to filter by. Also, not pictured in the mock ups(which i'll ask for a design for) is that there will be some filters that are drop downs instead of checkboxes. In general you can assume the following:

1. react-mui-table will be passed an array of objects for filters.
2. Each object in filters will contain an 'options' array. The shape of this array needs to be determined still.
3. Each object will have a 'mode' key that is either checkboxes, or dropdown. 

The design also has the label of the filter listed above the options. 

I'm fine with pushing this particular piece to a future PR as it will take a considerable rework of logic.

brad-decker commented on src/helpers/chunkArray.js:
```diff
@@ -0,0 +1,24 @@
+  return chunks;
+};
+
+export default chunkArray;
```
Lets export this as a reselect selector that calls the chunkArray function.

```js
export default createSelector(
  allItems => allItems,
  (_, subItemCount) => subItemCount,
  (allItems, subItemCount) => chunkArray(allItems, subItemCount),
);
```

kristajg replied:
Ok, this makes sense except for the underscore on line 3, not sure what that parameter means?

brad-decker commented on src/components/list/ListColumns.js:
```diff
@@ -57,17 +50,43 @@ export default class ListColumns extends Component {
+            <div
+              key={column.key}
+              style={styles.column}
+              onClick={column.sortable ? () => this.cycleSortOption(column.label) : NOOP}>
```
```js
{columns.map(column => (
 <div
    key={column.key}
    id={column.label}
    data-sortable={column.sortable}
    onClick={this.cycleSortOption}>
      {` ${generateSortPrefix(this.state[column.label])} ${column.label} `}
  </div>
)}
```

then in cycleSortOption:

```js
cycleSortOption(event) {
  if (!event.target.getAttribute('data-sortable')) return; // Don't do anything if not sortable. 
  const label = event.target.id;
  if (!Array.isArray(this.state[label])) return;
  const slicedArr = this.state[label].slice();
  slicedArr.push(slicedArr.shift());
  this.setState(
    { [label]: slicedArr },
    () => this.props.handleSort(label, this.state[label][0])
  );
}
```

Nathan-Schwartz replied:
I like this approach 👍

brad-decker commented:
We'll do squashing once review is approved :+1: @kristajg

martahj commented on example/app.js:
```diff
@@ -1,7 +1,8 @@
 import injectTapEventPlugin from 'react-tap-event-plugin';
-import List from '../dist/list/index.js';
+// import List from '../dist/components/list/index.js';
+import List from '../src/components/list';
```
Looks like there's a piece of commented out legacy code here - can we remove that?

kristajg replied:
Yep that's my bad. Was pointing List to src/components while working

martahj commented on src/components/list/ListAction.js:
```diff
@@ -4,9 +4,16 @@ import MenuItem from 'material-ui/MenuItem';
+  right: '10px',
+  top: '10px',
+};
+
```
Eventually it may be useful for the open source platform to also let the package user pass down optional styles as an override (at least of the right and top properties) or to provide additional styling - I don't see an urgent need right now though

martahj replied:
This comment goes for most of the style objects throughout the component (similar to how MUI gives you the option of overriding most of its default styling)

kristajg replied:
@martahj I agree, definitely will be a nice to have moving forward

martahj commented on src/components/list/ListRow.js:
```diff
@@ -50,10 +63,12 @@ export default class ListRow extends Component {
           <Checkbox
+            style={styles.checkbox}
             onCheck={this.createItemCheckHandler(item)}
             checked={Boolean(this.props.deepFind(item, this.props.itemsSelected))} />
```
So we don't have to call this.createItemCheckHandler in render, what if we had:

```
class ListRow {
  itemCheckHandler = (event, checked) => {
    const { item } = this.props;
    if (checked) {
      this.props.checkRow(item);
    } else {
      this.props.uncheckRow(item);
    }
  }

  render() {
    return (
      <Checkbox onCheck={this.itemCheckHandler} />
    )
  }
}
```

martahj commented on src/components/masthead/Delete.js:
```diff
@@ -2,15 +2,32 @@ import React from 'react';
+    </ListItem>
+  );
+};
 
```
Why remove the destructuring?

martahj commented on src/components/search/index.js:
```diff
@@ -4,14 +4,35 @@ import SearchIcon from 'material-ui/svg-icons/action/search';
 // TODO: Debounce / throttle search
+
 const Search = props =>
-  <Paper zDepth={2} style={styles}>
```
props destructuring would be cool here too

martahj commented on src/components/masthead/Delete.js:
```diff
@@ -2,15 +2,32 @@ import React from 'react';
+      <DeleteIcon
+        style={styles.iconStyle}
+        onClick={props.handleDelete(props.itemsSelected)} />
+    </ListItem>
```
Assuming handleDelete is a function and not a function factory passing it into onClick like this will invoke the function as soon as the component is rendered. This can be avoided by passing in a function that is not immediately called:

const Delete = props => {
  const handleItemDelete = () => {
    props.handleDelete(props.itemSelected);
  }
  return (
    <ListItem
      style={styles.listStyle}>
      {`${props.itemsSelected.length} item${props.itemsSelected.length <= 1 ? '' : 's'} selected`}
      <DeleteIcon
        style={styles.iconStyle}
        onClick={handleItemDelete} />
    </ListItem>
  );
};

brad-decker commented on src/components/search/index.js:
```diff
@@ -4,16 +4,41 @@ import SearchIcon from 'material-ui/svg-icons/action/search';
+      style={styles.searchBar}>
+      <SearchIcon
+        style={styles.icon} />
+      <TextField
```
@nataliaisabelle @iamdavid0091 the hint text default for material-ui starts with darkBlack (0,0,0,0.87) and then does a fade function on it with a parameter of (.3). I'm not sure what the resulting color is but its different than from the mockup i think.

If its possible could you make a note of this? If we think the hint text styling color should always be the value we decide we could start building a material-ui theme that we use as a base instead of the default so we are doing less overrides on a component level. We could also support the user passing the theme object into our react-mui-table component to provide their themed coloring to the table.

brad-decker commented on src/components/pagination/index.js:
```diff
@@ -5,22 +5,51 @@ import ArrowRight from 'material-ui/svg-icons/hardware/keyboard-arrow-right';
+    position: 'relative',
+    top: 5,
+    float: 'right',
+    color: 'rgb(158, 158, 158)',
```
Make sure you are referencing the measurement docs for appropriate colors. This one is not a match, and its used quite frequently through out the code base.

brad-decker commented on src/components/pagination/index.js:
```diff
@@ -5,22 +5,51 @@ import ArrowRight from 'material-ui/svg-icons/hardware/keyboard-arrow-right';
+    <ListItem
+      disabled={true}
+      style={styles.wrapper}>
+      <div style={styles.pagination}>
```
You're using this div for both positioning and aesthetics, which in general is fine, but some of the aesthetics will be inherited by the children, such as the color, fontSize and fontWeight. Make sure this is the intended behavior.

kristajg replied:
@brad-decker Yes, that was my original intention, but for the material-ui specific components, they don't seem to inherit styles by default (it may depend on the mui component) but at least for icons & the dropdown, they need specific styles as far as I can tell.

brad-decker commented on src/components/list/ListColumns.js:
```diff
@@ -57,17 +43,45 @@ export default class ListColumns extends Component {
+              id={column.label}
+              data-sortable={column.sortable}
+              onClick={this.cycleSortOption}>
+              {` ${generateSortPrefix(this.state[column.label])} ${column.label} `}
```
The mock ups show an arrow pointing up or down instead of the words ASC or DESC.

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -43,29 +70,39 @@ class Filter extends Component {
-                </div>
-                )
+      <span>
+        <ListItem
```
The filter/delete/selections bar at the top of the table should not be a ListItem. It should be a div.

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -43,29 +70,39 @@ class Filter extends Component {
+      <span>
+        <ListItem
+          style={styles.filterBar}
+          primaryText='FILTER'
```
Instead of using a ListItem with primaryText, please use a FlatButton that is inside our header div.

brad-decker commented on src/components/list/ListColumns.js:
```diff
@@ -57,17 +43,45 @@ export default class ListColumns extends Component {
+          {columns.map(column =>
+            <div
+              key={column.key}
+              style={styles.column}
```
Any sortable column needs to have a pointer cursor that shows that it is clickable.

Nathan-Schwartz commented on src/components/list/index.js:
```diff
@@ -72,52 +72,63 @@ export default class ReactMuiTable extends Component {
+                  )}
+                <Pagination
+                  hasNextPage={typeof this.props.hasNextPage === 'boolean' || warning('hasNextPage bool')}
+                  hasPreviousPage={typeof this.props.hasPreviousPage === 'boolean' || warning('hasPreviousPage bool')}
```
Checking if the value is a boolean is a good idea, but I don't think it should be done here. This code is replacing the value of hasNextPage with `true` if given `false` or `false` if passed a non-boolean truthy value.

Prop validation also doesn't need to be done on each render. I would recommend extracting all of this validation code into a validation function that gets invoked in the constructor and componentWillReceiveProps. 

Also note that we might not need to warn a user if filter/sort/etc options aren't passed in. Some people may not want to take advantage of all of the tables capabilities.

kristajg replied:
@Nathan-Schwartz Got it, sounds like we have many props that are required and some that are optional.  I agree this isn't the right place to be checking for existence / types.

I can make a validation function that could run on initial componentWillMount (unless I think of a better way to do this, maybe simply outside the class) and throws errors for missing required props.  For non-required props that are included without options being passed, I can have an abstracted warning function run, as needed.

kristajg replied:
@Nathan-Schwartz I used PropTypes, but noticed our linter threw a bunch of warnings.  Would you prefer I bring in flow types?

Nathan-Schwartz replied:
Yes, that would be awesome. 👍

Nathan-Schwartz commented on src/components/masthead/index.js:
```diff
@@ -4,14 +4,19 @@ import Delete from './Delete';
 const Masthead = (props) => {
-  if (props.itemsSelected.length) {
-    return <Delete itemsSelected={props.itemsSelected} handleDelete={props.handleDelete} />;
+  const { itemSelectedCount, handleDeletePointer, filters, handleFilter } = props;
```
Destructuring the props in the parameters would save us a line here.

Nathan-Schwartz commented on src/components/search/index.js:
```diff
@@ -4,16 +4,41 @@ import SearchIcon from 'material-ui/svg-icons/action/search';
-      onChange={e => props.handleSearch(e.target.value)} />
-  </Paper>;
+const Search = (props) => {
+  const { tableName, handleSearch } = props;
```
Same here

Nathan-Schwartz commented on src/components/pagination/index.js:
```diff
@@ -5,22 +5,51 @@ import ArrowRight from 'material-ui/svg-icons/hardware/keyboard-arrow-right';
+};
+
 const Pagination = (props) => {
   const { hasNextPage, hasPreviousPage, numRows } = props;
```
Destructuring would be nice here as well

Nathan-Schwartz commented on src/components/masthead/Delete.js:
```diff
@@ -2,15 +2,34 @@ import React from 'react';
-  </ListItem>
-);
+const Delete = props => {
+  const { itemSelectedCount, handleDeletePointer } = props;
```
Is there a reason to remove destructuring from the parameters here?

Nathan-Schwartz commented on src/components/list/ListColumns.js:
```diff
@@ -57,17 +57,48 @@ export default class ListColumns extends Component {
+      },
+    };
+
+    console.log('columns ', columns);
```
Leftover console.log

Nathan-Schwartz commented on src/components/list/ListAvatar.js:
```diff
@@ -14,7 +15,7 @@ const ListAvatar = (props) => {
   return (
     <Avatar
-      style={style}
+      style={{ ...styles, top: !src ? '' : 5 }}
```
I think it would be a little cleaner to do something like this so that top isn't assigned a blank string.

`const topStyle = src ? {top: 5} : {};`

`style={{ ...styles, ...topStyle }}`

kristajg commented on src/components/pagination/index.js:
```diff
@@ -5,22 +5,51 @@ import ArrowRight from 'material-ui/svg-icons/hardware/keyboard-arrow-right';
-    <ListItem>
-      <div>
+    <ListItem
+      disabled={true}
```
Ok agreed, I will swap out all ListItem components in favor of divs for more styling control

brad-decker replied:
Sorry @kristajg was off the grid for a bit due to the move. I wasn't suggested we change ALL the list items to div .Just the header and footer. The individual item rows should be list items because we do want the ripple (i think, might wanna get @iamdavid0091 and @nataliaisabelle 's feedback on this. If they don't want the ripple at all on the list items on click evens we can keep these as Divs)

brad-decker commented on src/components/pagination/index.js:
```diff
@@ -1,29 +1,56 @@
-          numRows={numRows} />
+      <div style={styles.caretWrapper}>
+        <IconButton
+          iconStyle={{ color: hasPreviousPage ? 'rgba(0, 0, 0, .54)' : 'rgba(0, 0, 0, .38)', ...styles.leftIcon }}
```
@iamdavid0091 @nataliaisabelle -> I think the padding between the buttons for pagination is wrong but i could be mistaken. Just in terms of clickable area. @kristajg has correctly positioned the arrows, but in order to be correct she will have to actually position the iconButton for the click action. If you look at this screenshot:

<img width="317" alt="screen shot 2016-12-31 at 12 52 15 pm" src="https://cloud.githubusercontent.com/assets/4448075/21579018/005a46e4-cf5e-11e6-8956-0eb4b26a657d.png">

you can see that the icon is positioned but the button for the icon is off to the left... but if we move the button itself it'll overlap the other button clickable area. Are you okay with us just positioning the buttons side by side with default material-ui padding/clickable area measurements?

brad-decker replied:
@kristajg - @iamdavid0091 and @nataliaisabelle are gonna be a bit busy today. Can you go ahead and just float those pagination buttons together without manipulating their position so we can get the click / ripple effect to be correctly positioned behind the button. We can figure out a new solution later if they think the spacing is too great.

kristajg replied:
Will do!

brad-decker commented on src/components/masthead/Filter.js:
```diff
@@ -43,29 +72,40 @@ class Filter extends Component {
-                </div>
-                )
+      <span>
+        <div
```
This is a div and should be a material-ui flat button.

kristajg replied:
@brad-decker Ok will do. Is it the intention here just to change the on-click styling (ripple etc)?

brad-decker replied:
Sort of. I mean it should be a button not just a clickable element. So beyond just having the right click effect/etc it is better syntax to use the Material-UI button element

kristajg commented:
@brad-decker Fixed the col width discrepancy, added flat button in filter bar, and fixed several other style flaws between avatar versus non-avatar tables.

kristajg commented:
@brad-decker ready for review

Nathan-Schwartz commented on src/components/list/ListRow.js:
```diff
@@ -1,43 +1,77 @@
+  };
+
+  itemCheckHandler = (event, checked) => {
+    const { item } = this.props;
```
If we are already destructuring, we may as well destructure `checkRow` and `uncheckRow`

Nathan-Schwartz commented on src/components/list/ListColumns.js:
```diff
@@ -57,17 +55,45 @@ export default class ListColumns extends Component {
-          <span
+      <div style={styles.wrapper}>
+        <Checkbox
+          style={{ ...styles.checkbox, paddingRight: this.props.avatar ? 64 : '' }}
```
We may not want to be setting the paddingRight to an empty string, while it may not alter styles I think it is confusing to read. How about using a pattern like so: 
```js 
const avatarStyles = this.props.avatar ? { top: 5 } : {};
```

Nathan-Schwartz commented on src/components/list/ListRow.js:
```diff
@@ -47,15 +81,18 @@ export default class ListRow extends Component {
-        </ListItem>
-        <Divider />
+      <div
+        style={{ ...styles.row, backgroundColor: this.state.hoverState ? '#eee' : '' }}
```
same thing as above 👍

Nathan-Schwartz commented on src/helpers/chunkArray.js:
```diff
@@ -0,0 +1,30 @@
+    const chunkValues = allItems.slice(startIndex, endIndex);
+
+    chunks.push(chunkValues);
+    // eslint-disable-next-line no-plusplus
```
Rather than disable the linting rule, you could do `currentChunk += 1`

Nathan-Schwartz commented on src/components/list/ListRow.js:
```diff
@@ -1,43 +1,77 @@
-  display: 'inline-block',
-};
-
 // eslint-disable-next-line react/prefer-stateless-function
```
If there is more than a render method present in a component the linter shouldn't complain. Could you check if disabling react/prefer-stateless-function is still necessary everywhere?

kristajg replied:
You are correct, removing the two instances I found.

brad-decker commented:
Squash dem commits again ( git rebase -i --autosquash origin/master ) and then i think this is mergable. :+1:

Nathan-Schwartz commented on src/components/list/ListRow.js:
```diff
@@ -47,16 +78,19 @@ export default class ListRow extends Component {
-        <Divider />
-      </div>
+      <ListItem
+        style={{ ...styles.row, backgroundColor: this.state.hoverState ? '#eee' : '' }}
```
This one looks like it snuck through.

Nathan-Schwartz commented on src/components/pagination/index.js:
```diff
@@ -1,29 +1,53 @@
-        </span>
+        <IconButton
+          iconStyle={{ color: hasNextPage ? 'rgba(0, 0, 0, .54)' : 'rgba(0, 0, 0, .38)' }}
+          onClick={hasNextPage ? nextPage : () => {}}>
```
I think this and the above `previousPage` functions could be optimized. The problem is that we are creating up to 2 new functions on each render. Here are 3 possible solutions.

This could be sidestepped by declaring a NOOP function outside the component and referencing it in both ternaries instead of creating an empty arrow function. 

Another school of thought would advocate creating a new function (something like `safeNextPage` which only invokes the `nextPage` prop when `hasNextPage` is true. `safeNextPage` would be passed to onClick instead of a ternary.

Or, you could assume that the consumer of the table has made sure that invoking `nextPage` when there is none, isn't harmful. `nextPage` is passed to onClick instead of a ternary.

What do you think? @brad-decker @kristajg

brad-decker replied:
onClick={this.nextPage}

```js
nextPage = () => {
if (!this.props.hasNextPage) return null;
this.props.nextPage();
}
```

kristajg replied:
@Nathan-Schwartz option one seems viable & straightforward.

Something like this above the component?

const NOOP = (pageBool, pageFunc) => {
  if(pageBool) {
    return pageFunc;
  }
}

then run NOOP inside the respective onClicks?

Nathan-Schwartz replied:
True, but it is actually a little more convoluted than number 2, which is what Brad is suggesting. 

NOOP stands for no operation, meaning it doesn't do anything. it would be `const NOOP = () => {}` and then the ternary would still be in place. Difference being that the `NOOP` would take the place of the anonymous arrow function. 

What you are suggesting is actually closer to number 2.

Try implementing what Brad has typed out. It could be made a method of either the pagination component, or the main component.

kristajg commented:
@Nathan-Schwartz Updated!

brad-decker commented:
can you amend your commit message to make it match the purpose the PR @kristajg ? Something like: Styling table to match designs ? This way once merged its clear what the intent of the commit was.

kristajg commented:
@brad-decker yes! done :)

Nathan-Schwartz commented on src/components/pagination/index.js:
```diff
@@ -1,28 +1,65 @@
+const Pagination = ({
+  hasNextPage, hasPreviousPage, numRows, paginationText, changeRowsPerPage, previousPage, nextPage,
+}) => {
+  const runNextPage = () => {
```
I could be mistaken here, but I believe that declaring these functions as local variables instead of class methods means that the they will still be created on each render. You could either make Pagination into an ES6 class or pull those method definitions into a parent component.

brad-decker replied:
Path of least resistance would be to make it into a class extending PureComponent

Nathan-Schwartz replied:
True, though I question using PureComponent here. We don't use it anywhere else in the package and using it could limit the compatibility of the package.

kristajg replied:
How about declaring the func above the component like @Nathan-Schwartz NOOP suggestion?  Any problems with that?

brad-decker replied:
The functions need access to this.props. It really needs to be a class here. If you do it that way you'd have to use .bind in the onClick call to pass props and that'd lead to the same problem of function creation.

Nathan-Schwartz replied:
It would require all of the data to be passed in, meaning the function would probably have to be wrapped in a new function in the render method.

brad-decker replied:
Jinx. You owe me a coke. @Nathan-Schwartz

Nathan-Schwartz replied:
😂 I didn't see your post, i'm going to blame it on lag

kristajg replied:
What about putting it in a react lifecycle method on a normal react component?

Nathan-Schwartz replied:
That is sort of what we are both suggesting, though I think we have different ideas about whether it should be a PureComponent or not. @brad-decker 

Difference being that we want it to be a class method, not a lifecycle method. Component lifecycle methods are class methods with certain names that React looks for.

Nathan-Schwartz replied:
@kristajg After some discussion Brad and I have agreed on using Component and using `isEqual` from lodash to do a props/state comparison in `shouldComponentUpdate`. It will probably be identical to what I did in the main list component. 👍

This way we get the benefits of PureComponent without reducing compatibility.

 Thanks for bearing with us!
